pkg/controllers/override/imageparse.go

```diff
@@ -0,0 +1,175 @@
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/distribution/reference"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	utilreference "github.com/kubewharf/kubeadmiral/pkg/lifted/distribution/reference"
+)
+
+type Image struct {
+	registry   string
+	repository string
+	tag        string
+	digest     string
+}
+
+func (i *Image) OperateRegistry(operator, value string) {
+	switch operator {
+	case OperatorAdd:
+		i.registry += value
+	case OperatorReplace:
+		i.registry = value
+	case OperatorRemove:
+		i.registry = ""
+	}
+}
+
+func (i *Image) OperateRepository(operator, value string) {
+	switch operator {
+	case OperatorAdd:
+		i.repository += value
+	case OperatorReplace:
+		i.repository = value
+	case OperatorRemove:
+		i.repository = ""
+	}
+}
+
+func (i *Image) OperateTag(operator, value string) {
+	var newTag string
+	switch operator {
+	case OperatorAdd:
+		newTag = i.tag + value
+	case OperatorReplace:
+		newTag = value
+	case OperatorRemove:
+		newTag = ""
+	}
+
+	if operator != OperatorRemove && !utilreference.AnchoredTagRegexp.MatchString(newTag) {
+		return
+	}
+
+	i.tag = newTag
+}
+
+func (i *Image) OperateDigest(operator, value string) {
+	var newDigest string
+	switch operator {
+	case OperatorAdd:
+		newDigest = i.digest + value
+	case OperatorReplace:
+		newDigest = value
+	case OperatorRemove:
+		newDigest = ""
+	}
+
+	if operator != OperatorRemove && !utilreference.AnchoredDigestRegexp.MatchString(newDigest) {
+		return
+	}
+
+	i.digest = newDigest
+}
+
+func (i *Image) String() string {
+	fullName := i.repository
+
+	if i.registry != "" {
+		fullName = strings.Join([]string{i.registry, i.repository}, "/")
+	}
+
+	if i.tag != "" {
+		fullName = strings.Join([]string{fullName, i.tag}, ":")
+	}
+
+	if i.digest != "" {
+		fullName = strings.Join([]string{fullName, i.digest}, "@")
+	}
+
+	return fullName
+}
+
+// ParseImage returns a Components of the given image.
+func ParseImage(fullName string) (*Image, error) {
+	ref, err := reference.Parse(fullName)
+	if err != nil {
+		return nil, err
+	}
+
+	image := &Image{}
+	if named, ok := ref.(reference.Named); ok {
+		image.registry, image.repository = SplitRegistryAndRepository(named.Name())
+	}
+
+	if tagged, ok := ref.(reference.Tagged); ok {
+		image.tag = tagged.Tag()
+	}
+
+	if digested, ok := ref.(reference.Digested); ok {
+		image.digest = digested.Digest().String()
+	}
+
+	return image, nil
+}
+
+func SplitRegistryAndRepository(name string) (registry, repository string) {
+	i := strings.IndexRune(name, '/')
+	if i == -1 || (!strings.ContainsAny(name[:i], ".:") && name[:i] != "localhost") {
+		registry, repository = "", name
+	} else {
+		registry, repository = name[:i], name[i+1:]
+	}
+	return
+}
+
+func ParseValueFromUnstructuredObj(input *unstructured.Unstructured, path string) (string, error) {
+	segments := strings.Split(strings.Trim(path, pathSeparator), pathSeparator)
+	currentObj := input.Object
+
+	for i := 0; i < len(segments)-1; i++ {
```

Comment:
Might want to refer to the implementation of `unstructured.NestedField`, which is more elegant imo. The main difference from the current implementation is to switch on `currentObj`'s type, instead of `currentObj[segment]`'s type. Thus, we don't need `segments[i+1]` and should be easier to reason.

Reply:
Is the path here supposed to be a JSON Patch [pointer](https://jsonpatch.com/#json-pointer)? If yes, is there an established package that we can use to do the parsing here?

Reply:
The code has been updated.